### PR TITLE
Bugfix/score matching equinox upgrade

### PR DESCRIPTION
**Pull request by gw265981**
### PR Type

- Bugfix
- Build related changes

### Description

- Small change to the ScoreMatching class to make it work with equinox>0.11.7.
- Updated version constraint for equinox as it no longer needs to bounded from above.

### How Has This Been Tested?

Unit tests now pass.

### Does this PR introduce a breaking change?

No.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [ ] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

File: coreax/score_matching.py
Comment (rg936672) [inline coreax/score_matching.py:63]: I don't get any test errors even without this change? (With the versions specified in the `uv.lock` on this branch.)

File: CHANGELOG.md
Comment (rg936672) [inline CHANGELOG.md:44]: This is technically a breaking change, and we should highlight it as such!

File: coreax/score_matching.py
Comment (gw265981) [reply coreax/score_matching.py:63]: Interesting, I think it fails if ABC is the first argument (like it was before) but not if it's the second one. I'm happy to add ABC as a second argument if there is a good reason for this?

File: CHANGELOG.md
Comment (gw265981) [reply CHANGELOG.md:44]: Is there a special format for this? Also, the contributor guidelines state that we need a deprecation period for breaking changes, but it seems like overkill in this instance.

Comment (rg936672) [reply CHANGELOG.md:44]: I don't think there's a specific format for it; [Keep a Changelog](https://keepachangelog.com/en/1.1.0/) doesn't specify one. I'd just add `**[BREAKING CHANGE]**` to the start of the line to make it obvious.

I agree that a deprecation period seems overkill here, especially since there's not really a way to have a transition period.

However, I'm not sure we actually need to limit the Equinox version so tightly in the first place - the tests seem to run fine with Equinox versions older than `0.11.9` (e.g. `0.11.7`)? We should definitely have _some_ lower bound on the Equinox version though - e.g. I'd imagine Coreax won't work with `equinox==0.0.1`!

File: coreax/score_matching.py
Comment (gw265981) [reply coreax/score_matching.py:63]: I was just following the example from equinox: https://docs.kidger.site/equinox/pattern/, who don't include ABC in the inheritance. I have tested that `@abstractmethod` is enough to enforce method implementation, I'm not sure if there is any other functionality we want. 

If it's just for the purpose of being explicit, I'm happy to include the explicit inheritance, although the decorator is fairly explicit. 

Anyway, this is a fairly insignificant change, so happy to go with your final call, but it's been a good discussion in case something like this crops up again.

File: CHANGELOG.md
Comment (gw265981) [reply CHANGELOG.md:44]: From my very rigorous testing, the tests fail for 0.11.0 but pass for 0.11.5, so the latter should be relatively safe. A potential issue is that the dev environment will often resolve to the latest version so we won't really know if something breaks for the older versions.

File: coreax/score_matching.py
Comment (rg936672) [reply coreax/score_matching.py:63]: Having looked a little closer into the code, `eqx.Module` is itself an abstract class (its metaclass is a subclass of `ABCMeta`) so we shouldn't need to inherit from ABC or explicitly set the metaclass ourself.

I think we'd want to do at least one of: 
- renaming it to `AbstractScoreMatching` (this one is what the Equinox docs do, but renaming is a fairly broad change which might be a hassle wrt deprecation periods etc)
- Making it very obvious in the docstring (in the first line ideally) that it's an _abstract_ base class

to make it immediately obvious that it's an abstract class.

File: CHANGELOG.md
Comment (rg936672) [reply CHANGELOG.md:44]: In which case, the latter is a good lower bound. Since tests fail for earlier versions, I think bumping the minimum required version isn't a breaking change _per se_, since it's just making a pre-existing requirement explicit, but it's still worth highlighting in the changelog.

Checking our lower bounds are correct is something that we can do by setting uv's [`--resolution-strategy`](https://docs.astral.sh/uv/concepts/resolution/#resolution-strategy) to `lowest` (or possibly `lowest-direct` in a pinch) - I'm planning on adding this in a CI job at some point, either as part of resolving #807 or as a subsequent PR.

**Review by rg936672 — APPROVED** [review]
Approved, assuming the CI passses.

**Pull request merged**
